### PR TITLE
Fixed all-deploys

### PR DESCRIPTION
**Pull request by nyarly**
The /all-deploy-queues endpoint was trying to serialize a map[sous.DeploymentID]int.
This was a problem because DeploymentID is a struct - which Go allows as map keys,
but JSON does not allow as object field names. There was the further issue that,
if simply converted to strings, clients would need to reconstruct the DeploymentID
to retrieve the specific queue.

Therefore, the structure of the response JSON changed. It returns an object keyed
by deploymentID strings, but where the values are objects with the full DeploymentID
included.

The change in the data structure meant that the tests needed to be overhauled,
and I wound up unrolling the complex table test driver.

To facilitate debugging, I refined the logging in the restful package to ensure
that we get logging of HTTP exchanges in some corner cases.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
LGTM

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: shell.nix [hunk 1]
```diff
@@ -1,5 +1,5 @@
-#{ pkgs ? import ~/dev/nixpkgs {} }:
-{ pkgs ? import <nixpkgs> {} }:
+{ pkgs ? import ~/dev/nixpkgs {} }:
+#{ pkgs ? import <nixpkgs> {} }:
```
Comment (JC1738) [inline shell.nix]: @nyarly Not sure I understand this change or what this file does?

File: util/restful/statusmiddleware.go [hunk 2]
```diff
@@ -45,22 +45,11 @@ func (ph *StatusMiddleware) errorBody(status int, rq *http.Request, w io.Writer,
...
-	if status >= 400 {
-		messages.ReportLogFieldsMessage("Status >= 400", logging.WarningLevel, ph.LogSink)
-		ph.errorBody(status, r, w, data, nil, nil)
-	}
-	if status >= 200 && status < 300 {
-		messages.ReportLogFieldsMessage("Status >= 200 && Status < 300", logging.DebugLevel, ph.LogSink)
-	}
-	// XXX in a dev mode, print the panic in the response body
-	// (normal ops it might leak secure data)
+	ph.errorBody(status, r, w, data, nil, nil)
```
Comment (JC1738) [inline util/restful/statusmiddleware.go]: Does this always suppose to call errorBody no matter what the status?

File: server/handle_alldeployqueues_test.go [hunk 3]
```diff
@@ -26,96 +26,98 @@ func TestNewAllDeployQueuesResource(t *testing.T) {
...
+func assertIsDeploymentQueuesResponse(t *testing.T, data interface{}) DeploymentQueuesResponse {
+	dr, ok := data.(DeploymentQueuesResponse)
+	if !ok {
+		t.Fatalf("got a %T; want a %T", data, dr)
+		return DeploymentQueuesResponse{}
+	}
+	return dr
+}
+
+func assertLength(t *testing.T, dr DeploymentQueuesResponse, wantLen int) {
```
Comment (samsalisbury) [inline server/handle_alldeployqueues_test.go]: "assertNumQueues" would be more specific.

File: server/handle_alldeployqueues_test.go [hunk 4]
```diff
@@ -26,96 +26,98 @@ func TestNewAllDeployQueuesResource(t *testing.T) {
...
+	gotLen := len(dr.Queues)
+	if gotLen != wantLen {
+		t.Fatalf("got %d queued deployments; want %d", gotLen, wantLen)
+	}
+}
+
+func assertQueueLength(t *testing.T, dr DeploymentQueuesResponse, did sous.DeploymentID, wantCount int) {
+	gotCount := dr.Queues[did.String()].Length
+	if gotCount != wantCount {
+		t.Errorf("got %d queued rectifications for %q; want %d", gotCount, did.String(), wantCount)
```
Comment (samsalisbury) [inline server/handle_alldeployqueues_test.go]: `%q` verb already calls `.String()` so we can drop that explicit call here.

File: server/handle_alldeployqueues_test.go [hunk 5]
```diff
@@ -26,96 +26,98 @@ func TestNewAllDeployQueuesResource(t *testing.T) {
...
+
+func assertSuccess(t *testing.T, status int) {
+	const wantStatusCode = 200
+	if status != wantStatusCode {
+		t.Errorf("got %d; want %d", status, wantStatusCode)
 	}
 
 }
 
+func assertIsDeploymentQueuesResponse(t *testing.T, data interface{}) DeploymentQueuesResponse {
```
Comment (samsalisbury) [inline server/handle_alldeployqueues_test.go]: Add `t.Helper()`

File: server/handle_alldeployqueues_test.go [hunk 6]
```diff
@@ -26,96 +26,98 @@ func TestNewAllDeployQueuesResource(t *testing.T) {
...
+}
+
+func assertLength(t *testing.T, dr DeploymentQueuesResponse, wantLen int) {
+	gotLen := len(dr.Queues)
+	if gotLen != wantLen {
+		t.Fatalf("got %d queued deployments; want %d", gotLen, wantLen)
+	}
+}
+
+func assertQueueLength(t *testing.T, dr DeploymentQueuesResponse, did sous.DeploymentID, wantCount int) {
```
Comment (samsalisbury) [inline server/handle_alldeployqueues_test.go]: Add `t.Helper()`

File: server/handle_alldeployqueues_test.go [hunk 7]
```diff
@@ -26,96 +26,98 @@ func TestNewAllDeployQueuesResource(t *testing.T) {
...
-
-			testCount := dr.Deployments[sous.DeploymentID{}]
-			if testCount != 0 {
-				t.Errorf("got %d for empty DeploymentID expected 0", testCount)
-			}
-		})
+	return handler.Exchange()
+}
+
+func assertSuccess(t *testing.T, status int) {
```
Comment (samsalisbury) [inline server/handle_alldeployqueues_test.go]: "assertStatusCode" would be more specific. Also add `t.Helper()` to get a more useful line number.

File: server/handle_alldeployqueues.go [hunk 8]
```diff
@@ -34,16 +34,14 @@ func (r *AllDeployQueuesResource) Get(_ http.ResponseWriter, _ *http.Request, _
...
+	data := DeploymentQueuesResponse{Queues: map[string]QueueDesc{}}
+
 	queues := h.QueueSet.Queues()
-	m := map[sous.DeploymentID]int{}
 	for did, q := range queues {
-		m[did] = q.Len()
+		data.Queues[did.String()] = QueueDesc{
+			DeploymentID: did,
+			Length:       q.Len(),
+		}
```
Comment (samsalisbury) [inline server/handle_alldeployqueues.go]: Nice, we should probably add links out to the individual queue endpoints as well as an enhancement; not part of this PR though.

File: test/server_integration_test.go [hunk 9]
```diff
@@ -194,6 +194,14 @@ func (suite integrationServerTests) TestUpdateStateDeployments_Update() {
 	suite.Len(data.Deployments, 3)
 }
 
+func (suite integrationServerTests) TestGetAllDeployQueues() {
+	data := server.DeploymentQueuesResponse{}
+	updater, err := suite.client.Retrieve("./all-deploy-queues", nil, &data, nil)
+	suite.NoError(err)
+	suite.Len(data.Queues, 0)
+	suite.NotNil(updater)
+}
```
Comment (samsalisbury) [inline test/server_integration_test.go]: Thanks for adding this, I'll see if I can expand this a bit to check for the non-zero case as well.

File: server/handle_alldeployqueues_test.go [hunk 10]
```diff
@@ -26,96 +26,98 @@ func TestNewAllDeployQueuesResource(t *testing.T) {
...
-		{
-			desc: "same DeploymentID twice",
-			dids: []sous.DeploymentID{newDid("one"), newDid("one")},
-			wantResult: deploymentsResponse{Deployments: map[sous.DeploymentID]int{
-				newDid("one"): 2,
-			}},
-		},
+}
+
+func setupExchange(t *testing.T, dids ...sous.DeploymentID) (interface{}, int) {
```
Comment (samsalisbury) [inline server/handle_alldeployqueues_test.go]: I'll rename this `setupAllDeployQueuesExchange` as there are a lot of methods named `Exchange`.  Also add `t.Helper()` for better failure line numbers.

File: util/restful/router.go [hunk 11]
```diff
@@ -129,45 +130,45 @@ func (mh *MetaHandler) genericHandling(resName string, factory ExchangeFactory,
 // GetHandling handles Get requests.
 func (mh *MetaHandler) GetHandling(resName string, factory ExchangeFactory) httprouter.Handle {
 	return func(rw http.ResponseWriter, r *http.Request, p httprouter.Params) {
-		rw, data, status := mh.genericHandling(resName, factory, rw, r, p)
-		rw.Header().Add("Access-Control-Allow-Origin", "*") //XXX configurable by app
-		mh.renderData(status, rw, r, data)
+		lrw, data, status := mh.genericHandling(resName, factory, rw, r, p)
```
Comment (samsalisbury) [inline util/restful/router.go]: Did this fix an issue, or just avoiding shadowing?

File: util/restful/statusmiddleware.go (hunk 2 above)
Comment (samsalisbury) [reply util/restful/statusmiddleware.go]: Agree this looks confusing, `errorBody` seems to write an error body no matter what the status as well.
